shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1272,65 +1284,220 @@ - (void)updateViewportPadding {
+  CGFloat keyboardBeginWidth = CGRectGetWidth(keyboardBeginFrame);
+  if (CGRectEqualToRect(keyboardEndFrame, CGRectZero)) {
+    return NO;
+  }
+  if (screenHeight == keyboardBeginWidth) {
+    return YES;
+  }
+  return NO;
+}
 
-  if (self.targetViewInsetBottom != 0) {
-    // Ensure the keyboard will be dismissed. Just like the keyboardWillChangeFrame,
-    // keyboardWillBeHidden is also in an animation block in iOS sdk, so we don't need to set the
-    // animation curve. Related issue: https://github.com/flutter/flutter/issues/99951
-    self.targetViewInsetBottom = 0;
-    NSTimeInterval duration =
-        [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
-    [self startKeyBoardAnimation:duration];
+- (FlutterKeyboardMode)calculateKeyboardAttachMode:(CGRect)keyboardFrame
+                                  notificationName:(NSNotificationName)notificationName {
+  // There are multiple types of keyboard: docked, undocked, split, split docked,
+  // floating, predictive-only, minimized. This function will categorize
+  // the keyboard as one of the following modes: docked, floating, or hidden.
+  // Docked mode includes docked, split docked, predictive-only (when opening via click),
+  // and minimized (when opened via click).
+  // Floating includes undocked, split, floating, predictive-only (when dragged and dropped),
+  // and minimized (when dragged and dropped).
+  if (notificationName == UIKeyboardWillHideNotification) {
+    return FlutterKeyboardModeHidden;
+  }
+
+  // If keyboard's dimensions and position are all zeroes,
+  // that means it's been dragged and therefore floating.
+  if (CGRectEqualToRect(keyboardFrame, CGRectZero)) {
+    return FlutterKeyboardModeFloating;
+  }
+  // If keyboard's width or height are 0, it's hidden.
+  if (CGRectIsEmpty(keyboardFrame)) {
+    return FlutterKeyboardModeHidden;
+  }
+
+  CGRect screenRect = [self getMainScreen].bounds;
+  CGFloat keyboardWidth = CGRectGetWidth(keyboardFrame);
+  CGFloat screenWidth = CGRectGetWidth(screenRect);
+
+  // If keyboard is not full width, it's floating.
+  if (keyboardWidth != screenWidth) {
+    return FlutterKeyboardModeFloating;
+  }
+
+  CGFloat screenHeight = CGRectGetHeight(screenRect);
+  CGRect adjustedKeyboardFrame = keyboardFrame;
+  adjustedKeyboardFrame.origin.y += [self calculateMultitaskingAdjustment:screenRect
+                                                            keyboardFrame:keyboardFrame];
+  CGFloat adjustedKeyboardBottom = CGRectGetMaxY(adjustedKeyboardFrame);
+
+  // If the keyboard is above the bottom of the screen, it's floating.
+  if (adjustedKeyboardBottom < screenHeight) {
+    return FlutterKeyboardModeFloating;
+  }
+  // If the keyboard is partially or fully showing at the bottom of the screen, it's docked.
+  if (CGRectIntersectsRect(adjustedKeyboardFrame, screenRect)) {
+    return FlutterKeyboardModeDocked;
+  }
```

Comment:
I think using `CGRectIntersectsRect(adjustedKeyboardFrame, screenRect)` is not very accurate. Maybe we should use some double epsilon to solve it. It will be `true` when keyboard is going to hidden **on some custom keyboard** (That what we fix in https://github.com/flutter/engine/pull/31982)

Reply:
Do you know if it's reproducible through a simulator? I tried creating a custom keyboard extension and then ran it in an iPhone 13 Pro iOS 15.4 simulator but was unable to reproduce

Reply:
I try to reproduce this on simulator, but can't reproduce.

**I use `Bitmoji` in app store to test this case, maybe you can try that on your physical phone😄.**

This can't be repro stably. I think maybe there are some bugs in custom keyboard app to cause this 843.99999999
Emm but I think in flutter engine side, we should have enough flexibility to cover that😄

Reply:
Thanks for pointing this out, I was able to get reproducing by explicitly giving my custom keyboard extensions a height of 397 + (2/3). 

Here's the solution I came up with, let me know if you have more feedback 🙂
https://github.com/flutter/engine/blob/6c4f65e1a322b861b8d664d64655b8e155f141c1/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L1404-L1418

Reply:
Yes, I think this will workaround ^_^

Reply:
@vashworth can you confirm then that https://github.com/flutter/flutter/issues/99951 hasn't regressed (per the link to https://github.com/flutter/engine/pull/31982 above)?  Might want to add custom keyboards to your list of manual tests in the description.

Reply:
I tested locally it will not regressed O(∩_∩)O